java-checks/src/main/java/org/sonar/java/checks/DateTimeFormatterMismatchCheck.java

```diff
@@ -0,0 +1,216 @@
+    .build();
+
+  private static final MethodMatchers WEEK_OF_WEEK_BASED_YEAR_MATCHER = MethodMatchers.create()
+    .ofTypes("java.time.temporal.WeekFields")
+    .names("weekOfWeekBasedYear")
+    .addParametersMatcher()
+    .build();
+
+  private static final Pattern WEEK_PATTERN = Pattern.compile(".*ww{1,2}.*");
+  private static final Pattern YEAR_OF_ERA_PATTERN = Pattern.compile(".*[uy]+.*");
+
+  private static final String CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE = "Change this year format to use the week-based year instead.";
+  private static final String CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE = "Change this year format to use ChronoField.YEAR instead.";
+  private static final String SECONDARY_LOCATION_MESSAGE = "";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (OF_PATTERN_MATCHER.matches(invocation)) {
+      visitPattern(invocation);
+    } else if (APPEND_VALUE_MATCHER.matches(invocation)) {
+      ChainVisitor visitor = new ChainVisitor();
+      invocation.accept(visitor);
+      if (!visitor.usesWeek || !visitor.usesYear) {
+        return;
+      }
+      if (visitor.usesWeekBasedYear && !visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_TO_CHRONOFIELD_MESSAGE, visitor.locations, null);
+      } else if (!visitor.usesWeekBasedYear && visitor.usesWeekOfWeekBasedYear) {
+        reportIssue(visitor.primary, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE, visitor.locations, null);
+      }
+    }
+  }
+
+  /**
+   * We currently only check formatters initialized with a String literal or the value of a final variable.
+   */
+  private void visitPattern(MethodInvocationTree invocation) {
+    Arguments arguments = invocation.arguments();
+    ExpressionTree argument = arguments.get(0);
+    if (argument.is(Tree.Kind.STRING_LITERAL)) {
+      String pattern = ((LiteralTree) argument).value();
+      if (isInfringingPattern(pattern)) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE);
+      }
+    } else if (argument.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) argument;
+      Symbol symbol = identifier.symbol();
+      if (!symbol.isFinal()) {
+        return;
+      }
+      Tree declaration = symbol.declaration();
+      if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+        return;
+      }
+      VariableTree variable = (VariableTree) declaration;
+      ExpressionTree initializer = variable.initializer();
+      if (initializer == null || !initializer.is(Tree.Kind.STRING_LITERAL)) {
+        return;
+      }
+      String pattern = ((LiteralTree) initializer).value();
+      if (isInfringingPattern(pattern)) {
+        reportIssue(invocation, CHANGE_YEAR_FORMAT_WEEK_BASED_MESSAGE);
+      }
```

Comment:
I believe it would be easier to read if we had less if/return statements by adding an helper method to retrieve the string value: ex: `getStringLiteralValue`